doc/source/architecture.rst

```diff
@@ -0,0 +1,213 @@
+
+Command pattern
+---------------
+
+Many interactive applications use the Command pattern. In a nutshell, anything
+the user does within the application turns into a Command object, and then the
+Command object is immediately executed. This approach carries some benefits.
+For example, applications that work in this way can choose to implement undo
+and redo by storing an 'inverse' function within the Command and storing a
+stack of executed commands. Undo/Redo is then implemented by walking up and
+down the stack and executing the Command or its inverse function.
+
+More relevant to this discussion is how command patterns can be used to
+implement automation APIs. If every user action is a Command, then that Command
+can serve as an API. In this way, everything the user does can be trivially
+recorded as textual representations of those Commands. These can then be
+replayed using a scripting language supported by the application. In fact, this
+exact command pattern is how the Ansys Electronics Desktop, Ansys Mechanical
+APDL, and the Ansys Workbench applications implement automation APIs for
+scripting.
+
+Using the command pattern as a scripting API does come with some disadvantages.
+Most importantly, command APIs are not symmetric. The four fundamental
+operations on data within a software application are Create, Read, Update, and
+Delete. These are often abbreviated as CRUD. While you can conceptually Update,
+Create, and Delete using Commands, you can not Read. This makes it difficult to
+"visit" the application's data model.
+
+Mechanical API implementation
+-----------------------------
+
+Mechanical's API serves the needs of both *automation* and *customization*. For
+customization, it is necessary to Read the data model. Automation alone is not
+sufficient. For example, when integrating a third-party solver, you will need
+to access all of the boundary conditions, geometry, material properties, mesh,
+and connections to properly input them to the solver. As discussed above, the
+command pattern does not allow this kind of access pattern. Due to this fact,
+the Mechanical API exposes its data model directly to the user. This is how a
+hypothetical command-based API would look for renaming an object.
+
+.. code:: python
+
+    RenameCommand(Id=100, Name="New name")
+
+Instead, an API based on a data model, like mechanical, looks like this:
+
+.. code:: python
+
+    obj = DataModel.GetObjectById(100)
+    print(obj.Name)
+    obj.Name = "New name"
+
+Notice that the old name could be printed by accessing a property of the
+object. This has no equivalent if the API used a command pattern.
+
+
+PyMechanical remote interface
+-----------------------------
+
+One way to interact with the Mechanical API from CPython is as a Remote
+Session. A session of Mechanical runs as a server and can receive a Remote
+Procedure Call (RPC). The RPC is handled by the server and then the response
+is returned back to the client.
```

Comment:
@PipKat  I have  habit of doing Some Term of Art (STA) rather than STA (Some Term of Art).

I see that you prefer the latter - is that a personal preference or recommended by our style guide?

Reply:
@koubaa My being wishy-washy on this is because the Google developer's doc style guide does say to use the first way:

    In general, when an abbreviation is likely to be unfamiliar to the audience, spell out the first 
    mention of the term and immediately follow with the abbreviation in parentheses, as in the 
    following examples:

    Border Gateway Protocol (BGP)
    elliptic-curve cryptography (ECC)

However, the trend in lots of Ansys materials and the materials that I see from lots of companies is to use the second way:

BGP (Border Gateway Protocol)
ECC (elliptic-curve cryptography)

At this point, within any particular document, I shoot for consistency. I'm OK with whichever your prefer.